test/distributed/test_ddp_with_rpc.py

```diff
@@ -0,0 +1,443 @@
+            self.process_group_for_ddp
+            if ddp_mode in (DdpMode.INSIDE,)
+            else None,
+        )
+        if ddp_mode == DdpMode.OUTSIDE:
+            gLogger.info(f"Wrapping the whole hybride module into DDP.")
+            self.hybrid_module = DistributedDataParallel(
+                self.hybrid_module,
+                process_group=self.process_group_for_ddp,
+                check_reduction=True,
+            )
+        gLogger.info(f"Succeeded in creating a HybridModel instance.")
+
+    def __del__(self):
+        dist.destroy_process_group(self.process_group_for_ddp)
+
+    def do_forward_without_grad(self, input: FeatureSet):
+        gLogger.debug(f"Doing a forward pass on {input}")
+        with torch.no_grad():
+            output = self.hybrid_module(input)
+            return self.criterion(output, input.labels)
+
+    def get_real_module(self):
+        if isinstance(self.hybrid_module, DistributedDataParallel):
+            return self.hybrid_module.module
+        else:
+            return self.hybrid_module
+
+    def print_parameters(self):
+        self.get_real_module().print_parameters()
+
+    def do_mini_batch(self, mini_batch: FeatureSet):
+        gLogger.info(f"Doing a mini batch on {mini_batch}")
+        loss = 0
+
+        with dist_autograd.context() as context_id:
+            output = self.hybrid_module.forward(mini_batch)
+            with torch.no_grad():
+                gLogger.debug(
+                    f"Output: {output} softmax: {torch.softmax(output, 1)} labels: {mini_batch.labels}"
+                )
+            loss = self.criterion(output, mini_batch.labels)
+            # grads will be stored in dist_autograd context
+            dist_autograd.backward(context_id, [loss])
+            gLogger.debug(
+                f"Distributed grads: {dist_autograd.get_gradients(context_id)}"
+            )
+            gLogger.info(f"Running distributed optimizer...")
+            dist_optimizer = DistributedOptimizer(
+                optim.SGD,
+                self.get_real_module().remote_parameters_rrefs
+                + [
+                    rpc.RRef(local_param)
+                    for local_param in self.hybrid_module.parameters()
+                ],
+                lr=LR,
+            )
+            dist_optimizer.step(context_id)
+
+            self.print_parameters()
+
+        gLogger.info(f"Loss: {loss}.")
+        return loss
+
+
+def get_training_examples():
+    n = 8
+    training_examples = FeatureSet(
+        dense_features=torch.zeros((n, D_DENSE)),
+        sparse_features=torch.zeros(n, dtype=torch.long),
+        labels=torch.zeros(n, dtype=torch.long),
+    )
+    idx = 0
+    for x in range(2):
+        for y in range(2):
+            for z in range(2):
+                training_examples.dense_features[idx, :] = torch.Tensor((x, y))
+                training_examples.sparse_features[idx] = z
+                training_examples.labels[idx] = x ^ y ^ z
```

Comment:
is there any reason for these value assignments?

Reply:
Not really. There are 8 examples as the 8 vertices of a cube in a 3-dimensional space. Each edge of the cube connects 2 vertices with different labels.